### ISSUE TITLE
Custom objective differs from default

### ISSUE THREAD

**Issue by martinju** [dmlc/xgboost#3255]
Working with custom objective functions, I realized that coding logistic regression myself gives a quite different result than using the default.

I am using R from Linux, xgboost 0.7.0 and coding custom objective functions based on this: 
https://github.com/dmlc/xgboost/blob/master/R-package/demo/custom_objective.R 

I have earlier experienced slightly different results using "hist" and "gpu_hist", which I believe is due to round-off errors, but I have not seen differences this large before. The differences on my very simple test below are more than a percent for all predictions. On other non-public data sets I got much larger differences.

Code to reproduce:
```
rm(list=ls())

library(xgboost)

invlogit <- function(x){
  1/(1 + exp(-x))
}

# load in the agaricus dataset
data(agaricus.train, package='xgboost')
data(agaricus.test, package='xgboost')
dtrain <- xgb.DMatrix(agaricus.train$data, label = agaricus.train$label)
dtest <- xgb.DMatrix(agaricus.test$data, label = agaricus.test$label)


logregobj <- function(preds, dtrain) {
  labels <- getinfo(dtrain, "label")
  preds <- 1/(1 + exp(-preds))
  grad <- preds - labels
  hess <- preds * (1 - preds)
  hess <- pmax(hess,10^(-16))
  return(list(grad = grad, hess = hess))
}

param.1 <- list(max_depth=2, eta=1, nthread = 2, silent=1, 
              objective=logregobj,tree_method="hist")

param.2 <- list(max_depth=2, eta=1, nthread = 2, silent=1, 
                objective="binary:logitraw",tree_method="hist")

bst.1 <- xgb.train(param.1, dtrain, 1)
bst.2 <- xgb.train(param.2, dtrain, 1)

invlogit(head(predict(bst.1,dtest)))
invlogit(head(predict(bst.2,dtest)))

```

It seems others have seen the same (https://github.com/dmlc/xgboost/issues/3096), but such large differences cannot due to round-off errors? Note that I have also used the trick used in the default code avoiding hessians which are too small.

Source code links I used: 
Logisticraw-code: https://github.com/dmlc/xgboost/blob/master/src/objective/regression_loss.h#L77
Sigmoid function: https://github.com/dmlc/xgboost/blob/master/src/common/math.h#L23

Am I doing something wrong?
If not, how can this happen? Is it a bug?
Any ideas are highly appreciated.

**Comment by tqchen** [comment]
The best way to find the problem would dig into the code and see if there is an actual difference in them, it would also be helpful to see if the default exact method is consistent.

@RAMitchell can you look into this?

**Comment by khotilov** [comment]
@martinju keep in mind that the default `base_score=0.5`, and there's no transformation-to-margin for it with the custom objective, while "logitraw" still has a plogis transformation-to-margin for the base_score:
https://github.com/dmlc/xgboost/blob/66e74d222330adf3e41bb1e1f34ae1dd4aae15f0/src/objective/regression_loss.h#L61
just like all the other LogisticRegression-based losses. Yes, it's confusing, and I tend not to use logitraw myself since I always forget how it behaves and have to looks at its code.
```r
> param.1 <- list(max_depth=2, eta=1, objective=logregobj)
> param.2 <- list(max_depth=2, eta=1, objective="binary:logitraw")
> bst.1 <- xgb.train(param.1, dtrain, 1)
> bst.2 <- xgb.train(param.2, dtrain, 1)
> head(predict(bst.1, dtest))
[1] -1.824148  1.800880 -1.824148 -1.824148 -2.084654  1.800880
> head(predict(bst.2, dtest))
[1] -1.700441  1.712177 -1.700441 -1.700441 -1.940709  1.712177
> 
> b = 0.
> param.1 <- list(max_depth=2, eta=1, objective=logregobj, base_score=b)
> param.2 <- list(max_depth=2, eta=1, objective="binary:logitraw", base_score=plogis(b))
> bst.1 <- xgb.train(param.1, dtrain, 1)
> bst.2 <- xgb.train(param.2, dtrain, 1)
> head(predict(bst.1, dtest))
[1] -1.700441  1.712177 -1.700441 -1.700441 -1.940709  1.712177
> head(predict(bst.2, dtest))
[1] -1.700441  1.712177 -1.700441 -1.700441 -1.940709  1.712177
 ```

**Comment by tqchen** [comment]
Thanks to @khotilov , this is resolved, further questions are encouraged at https://discuss.xgboost.ai/
